### DISCUSSION TITLE
Managing options and user preferences with a better GUI

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#2472] [Ideas]
Here are some ideas for managing options and user preferences with a better GUI.
Images are for explanation only.

"I apologize for the crudity of this model." - Emmett "Doc" Brown. :D

**Options Home screen.**

This screen allows to change the options settings.

![Options Home screen](https://user-images.githubusercontent.com/43583994/103087898-51fa8080-45e9-11eb-8cce-be65dab8eab8.png)

The buttons "Game", "World", "Heroes", "Battle", "Castle" and "Unions" represent the categories of the options.
By pressing one of these buttons, only the options related to the category are displayed.
The buttons have two states, selected and unselected.
Several buttons can be in a selected state at the same time (this allows to display several categories of options at the same time).
The blue arrow buttons, to the right of the "Unions" button, allow to display other possible categories.
The scroll bar allows to navigate through the options.

Several options can be combined in an option set.
An option can only be included in one option set.
An option set can be included in another option set.
If there is an option from a category, for example from the "Game" category, in an option set, the option set is displayed when the user uses the "Game" button.

The "Display All Options" button displays all options and option sets (this resets all buttons of the categories to their unselected state).

The bar with the name "Default" indicates the name of the preferences file, "Default" being the default setting for all options in the game (the first time the game is launched or if the options have not been changed (option enabled/not enabled).

The "Personalized Options" button displays only the options or option sets selected as "Personalized" by the user.
This allows the user, for example, to have quick access to options and option sets that he often changes.

The "Display Only Sets" button displays only option sets.

The "Load Preferences" button allows to load a backup file with the status of all options and option sets.
All options and option sets are automatically updated when the preferences file is loaded.
The name bar ("Default" originally) is updated with the name of the preferences file.

The "Save Preferences" button saves the status of all options and option sets to a backup file.
This allows the user to save different game preferences and to be able to quickly change the configuration of all options and option sets.
The default name of the backup file is "User 01", then "User 02" if "User 01 already exists, etc.".
The user can give specific names to the different backups, for example "Masochist" for a very hard game, "Cool Game" for a quiet game, "Let's the fun begin", etc.

Saving the preferences in a backup file also allows, by the way, to create custom "mods".

"After version 1.0, I plan to make a mod for fheroes2 with small changes here and there, and add new secondary skills, new artifacts, new spells, etc.
Everything will be optional and once implemented in fheroes2, it will be possible to integrate its changes in other mods according to the players' wishes. :joy:"

Thus, once a whole bunch of ideas have been implemented in fheroes2, the user can create mods according to his wishes and share them with other users.
The other users can adjust the mods according to their wishes.

The "Edit Option Set" button opens another screen where the user can change the status of the different options in the option set (change settings).

The "Create Option Set" button opens another screen where the user can assemble options to create an option set (put several options in an option set).
The user can, for example, put together all options (and option sets) that he will never change in an option set, etc.

The "Default" button resets all options and option sets to their original state.

The "Show/Hide" button shows/hides all options and option sets according to the user setting.

The "Okay" button allows to save all changes and exits the screen.

The "Cancel" button allows to exit the screen without saving the changes.

The blue rectangles represent options and option sets.
- Within each rectangle there is an area (top left) where a graphic can be displayed.
This can be a graphic related to the category (default) or a graphic customized by the creator of the option or by the user.
- For each option or option set there is a name and a description.
- The location in the top right corner, where the green graphic with the tick mark is located, allows to activate/deactivate the option or option set.
- The location at the bottom left, where the green graphic with the eye (and the text "Hide") is located, allows hiding/showing the option or option set.
The graphic changes to a red graphic with the eye crossed out (and the text "Show") to indicate to the user that the option or option set is hidden (the "Show/Hide" button allows to show/hide the hidden options and option sets so the user can change their visibility state).
This allows the user to make the settings of his choice for several options and option sets once and for all and no longer display them in the interface using the setting hide/show.
- The location at the bottom center, where the green graphic is located with the thumb hand up (and the text "Personalized") allows to set the status of the option or option set so it can be displayed only when the user uses the "Personalized Options" button.

The first blue rectangle of the second line, is an option set.
- The option set allows to combine options.
For example, it is possible to combine three options related to the hero's movement into one option set.
This has several advantages.
First of all, the options that affect the behavior of the game in relation to the hero's movement are combined in an option set, which makes it easier to access the settings.
Secondly, it takes up less space in the interface, because it is possible to combine dozens of options in one option set.
Example with the option set "Hero Movement" which contains three options:
"Hero Movement"
Optimized Hero Movement (https://github.com/ihhub/fheroes2/issues/1214)
Hero Path And Fog Of War (https://github.com/ihhub/fheroes2/issues/1217)
Hero Stops Upon Discovery (https://github.com/ihhub/fheroes2/issues/1219)
- The second location in the top right corner, where the red graphic with the trashcan is located, allows to delete an option set.
The options contained in the option set then become/rebecome independent options.
- The number at the bottom right indicates the number of options contained in the option set.
- The position at the bottom right, where the green graph with the plus sign is located, opens another screen where the user can change the status of the different options in the option set (change settings)(same action as the "Edit Option Set" button).

**Editing screen for option sets.**

This screen allows to modify the status of the options contained in the option sets.

![Edit Set screen](https://user-images.githubusercontent.com/43583994/103087992-a1d94780-45e9-11eb-9347-081b6a794f34.png)

The bar with the name "Title - (Set)" indicates the name of the open option set.

The "Save Option Set" button allows to save all changes and exits the screen.

The "Cancel Set Changes" button allows to exit the screen without saving the changes.

The first blue rectangle in the upper left corner represents the currently open option set.
- The bottom right location, where the red graphic with the minus sign is located, allows to exit the screen without saving the changes (same action as the "Cancel Set Changes" button).

The rectangles in the second row represent the options contained in the currently open option set.
- The second location at the top right, where the red graphic with the trashcan is located, allows to remove the option contained in the currently open option set.
The removed option then becomes/rebecomes an independent option.

**Screen for creating option sets.**

This screen allows to create option sets.

![Create Set screen](https://user-images.githubusercontent.com/43583994/103088002-aa318280-45e9-11eb-8ec4-bf8379714795.png)

The user can create a new option set by dragging and dropping options and option sets from the "Choose the options..." columns to the "Set Being Created" column.

The two scrollbars are used to navigate through the options.

The "Save Created Set" button allows to save the newly created option set and to exit the screen.

The "Cancel Set Creation" button allows to exit the screen without saving the option set.

Ideally, the screen structure should be stored in a description file (text or other) that would allow the layout of the interface elements to be changed easily, the graphics used, the texts, their positions in the interface, etc., etc., (provided that certain well-defined rules are respected).
Each option has a unique number, and it is possible to display the options in a certain order in the interface (in the description file).
The interface description file also contains all the default values of the options.
The default values of the options are also hardcoded in the program in case the file is deleted by the user.

If you have read so far, Merry Christmas and Happy New Year to all. :christmas_tree:

**Comment by ihhub** [comment]
Hi @LeHerosInconnu , please take into consideration that even till now we do not have any designer to create a new content for us so even adding a new button with a different styling might be troublesome.

**Reply by LeHerosInconnu** [reply]
Hello @ihhub,

> Hi @LeHerosInconnu , please take into consideration that even till now we do not have any designer to create a new content for us so even adding a new button with a different styling might be troublesome.

All images were intended primarily for explanation. :)

Everything can be done with the graphics already present in the game.
It's even better because it allows to stay in the atmosphere of the game.

Image is still for explanation only. :)

![Heroes 2 Options fheroes2 01](https://user-images.githubusercontent.com/43583994/103231125-5b605180-4937-11eb-9cfe-a7520f30842c.png)

**Comment by Branikolog** [comment]
Good idea and presentation, @LeHerosInconnu!
After 1.0 release we'd definitely need some kind of sorting of options we'' implement in game.
I don't want to hurt anybody, but I want this improvement look much less itemized and more user-friendly, than I saw in ERA mod of H3. :)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Good idea and presentation, @LeHerosInconnu!
> After 1.0 release we'd definitely need some kind of sorting of options we'' implement in game.
> I don't want to hurt anybody, but I want this improvement look much less itemized and more user-friendly, than I saw in ERA mod of H3. :)

Thanks, I already have some improvements/changes to suggest. :)
